### DISCUSSION TITLE
ERROR : RuntimeError: getVariable: no variable named 'YoungModulus'

### DISCUSSION THREAD

**Discussion by RokyC** [thelfer/MFrontGenericInterfaceSupport#104] [Q&A]
When I was referring to [XX calculation](https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html), the following problems occurred. May I ask what might be the cause?

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In [8], line 16
     14 H = E*Et/(E-Et)
     15 for s in [m.s0, m.s1]:
---> 16     mgis_bv.setMaterialProperty(s, 'YoungModulus', E)
     17     mgis_bv.setMaterialProperty(s, "PoissonRatio", nu)
     18     mgis_bv.setMaterialProperty(s, "HardeningSlope", H)

RuntimeError: getVariable: no variable named 'YoungModulus'

**Comment by thelfer** [comment]
Would you just print the list of material properties expected by the behaviour.

```python
for mp in m.b:
   print("- {}".format(mp.name)) 
```

**Reply by RokyC** [reply]
Code as follows:
# modelling hypothesis
h = mgis_bv.Hypothesis.PlaneStrain
# loading the behaviour        
b = mgis_bv.load('src/libBehaviour.so','IsotropicLinearHardeningPlasticity',h)
# material data manager
m = mgis_bv.MaterialDataManager(b,ngauss)
for mp in m.b:
   print("- {}".format(mp.name)) 
for s in [m.s0, m.s1]:
    mgis_bv.setMaterialProperty(s, "YoungModulus", 150e9)
    mgis_bv.setMaterialProperty(s, "PoissonRatio", 0.3)
    mgis_bv.setMaterialProperty(s, "HardeningSlope", 150e6)
    mgis_bv.setMaterialProperty(s, "YieldStrength", 200e6)
    mgis_bv.setExternalStateVariable(s, "Temperature", 293.15)

I got this(The command had been carried out:
mfront --obuild --interface=generic IsotropicLinearHardeningPlasticity.mfront
Treating target : all
The following library has been built :
- libBehaviour.so :  IsotropicLinearHardeningPlasticity_AxisymmetricalGeneralisedPlaneStrain IsotropicLinearHardeningPlasticity_Axisymmetrical IsotropicLinearHardeningPlasticity_PlaneStrain IsotropicLinearHardeningPlasticity_GeneralisedPlaneStrain IsotropicLinearHardeningPlasticity_Tridimensional):

The error:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In [13], line 7
      5 # material data manager
      6 m = mgis_bv.MaterialDataManager(b,int(ngauss))
----> 7 for mp in m.b:
      8    print("- {}".format(mp.name)) 
      9 for s in [m.s0, m.s1]:

AttributeError: 'MaterialDataManager' object has no attribute 'b'

**Reply by thelfer** [reply]
Sorry. Just type:

```python
for mp in b:
   print("- {}".format(mp.name))
```

**Reply by RokyC** [reply]
Umm, I got this:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [7], line 7
      5 # material data manager
      6 m = mgis_bv.MaterialDataManager(b, int(ngauss))
----> 7 for mp in b:
      8    print("- {}".format(mp.name))
      9 for s in [m.s0, m.s1]:

TypeError: 'Behaviour' object is not iterable

**Comment by thelfer** [comment]
Sorry, I am doing 10 things at the same time and I am not testing before posting. Let us try this:

```cxx
for mp in b.material_properties:
   print("- {}".format(mp.name))
```

**Reply by RokyC** [reply]
Yes, the b.material_properties has nothing.
![image](https://user-images.githubusercontent.com/59686187/197535533-99a4cb90-5309-47e8-b81b-eb9c8b37a6ca.png)

**Comment by thelfer** [comment]
So you just compiler an implementation that does not require you to define any material property !

**Reply by RokyC** [reply]
Thank you very much for your patience. My computer programming ability is poor. Could you please tell me the specific implementation method? I am using python, should I add the variable I need through the method 'append' of material_properties?

**Reply by thelfer** [reply]
It is the other way around: you loaded an MFront behaviour which exports a list of material properties and external state variables that must be defined on the `MGIS` side to integrate the behaviour. Don't define what is not needed (at let MGIS complain if you forgot something)

**Reply by RokyC** [reply]
![image](https://user-images.githubusercontent.com/59686187/197543249-dd0d36fd-b58f-49b8-9069-1a3ff1d71518.png)
Is that the definition? However, the original mfront file has been defined and does not seem to load these state names.

**Reply by thelfer** [reply]
Yes. It is.

**Reply by RokyC** [reply]
But the previous problem was compiling with this mfront file.

**Reply by thelfer** [reply]
Are you sure ? Maybe you shall run the following simple tests to see if it works

[test-mgis.zip](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/9852300/test-mgis.zip)

**Reply by thelfer** [reply]
Can't say. Make a diff of both `MFront` files. Then you can remove the `include` and `src` directories and recompile the behaviour...

**Reply by RokyC** [reply]
I know the reason, I used the mfront file of the official website before (there is no material attribute definition), but after using the new mfront file, my jupyterlab kernel did not restart.I apologize for the delay caused by my mistake.

**Reply by thelfer** [reply]
No problem. One must make (a lot of) mistake(s) to learn. Good luck !

**Reply by RokyC** [reply]
Thank you very much!
